Add rocBLAS backend (#144)
oneapi-src/oneMKL: README.md

    </thead>
    <tbody>
        <tr>

>             <td rowspan=6 align="center">oneMKL interface</td>
>             <td rowspan=6 align="center">oneMKL selector</td>

Reviewer: Need one more row here to align the table
```suggestion
            <td rowspan=7 align="center">oneMKL interface</td>
            <td rowspan=7 align="center">oneMKL selector</td>
```

    </tbody>
</table>

### Supported Configurations with hipSYCL:

Supported domains: BLAS

#### Linux*

<table>
<thead>
<tr  align="center">
<th >Domain</th>
<th >Backend</th>
<th >Library</th>
<th >Supported Link Type</th>
</tr>
</thead>
<tbody>
<tr >
<td  rowspan="4" align="center">BLAS</td>
<td  align="center">x86 CPU</td>
<td  align="center">Intel(R) oneAPI Math Kernel Library</td>
<td  align="center">Dynamic, Static</td>
</tr>
<tr >
<td align="center">AMD GPU</td>

> <td align="center">rocBLAS </td>

Reviewer 2: ```suggestion
<td align="center">AMD rocBLAS </td>
```

<td align="center">Dynamic, Static</td>
</tr>
<tr >
<td  align="center">NVIDIA GPU</td>
<td  align="center">NVIDIA cuBLAS</td>
<td  align="center">Dynamic, Static</td>
</tr>
<tr >
<td  align="center">x86 CPU</td>
<td  align="center">NETLIB LAPACK</td>
<td  align="center">Dynamic, Static</td>
</tr>
</tbody>
</table>

> #### Windows*

Reviewer 2: I would recommend removing Windows section and add it when it passes the experimental stage.

    </tbody>
</table>

> ### Supported Configurations with hipSYCL:

Reviewer: Could you please add "For compiling with DPC++ compiler" for the table above? So that it will be easier to see what compiler supports what.

Author: I am not sure if this should be part of this PR, but shouldn't the intel LLVM compiler and the dpc++ compiler be handled differently here?

Reviewer: Thanks!

Reviewer: Agree, we do have a difference in support for dpc++ and llvm compilers that we didn't reflect here. Maybe to avoid the duplication in the multiple tables what do you think about adding one more column `Supported compilers`  and have only one table? So e.g. for `Nvidia cuBLAS`  we can list LLVM and hipSYCL compilers in this column.

Reviewer 2: We have not tested cuBLAS backend with hipSYCL locally yet. Could you save it for a future PR?

Author: I also added an extra table to clarify the current domain support

Reviewer: Thanks for tables in the build section!
For this one I still think it will be easier to read as one table instead of two. For some reasons I can't open PR with the suggested table to your branch.

So here is the table in the row format

```
### Supported Configurations:

Supported domains: BLAS, LAPACK, RNG

#### Linux*

<table>
    <thead>
        <tr align="center" >
            <th>Domain</th>
            <th>Backend</th>
            <th>Library</th>
            <th>Supported Link Type</th>
            <th>Supported Compiler</th>		
        </tr>
    </thead>
    <tbody>
        <tr>
            <td rowspan=5 align="center">BLAS</td>
            <td align="center">x86 CPU</td>
            <td rowspan=2 align="center">Intel(R) oneAPI Math Kernel Library</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*, hipSYCL</td>
        </tr>
        <tr>
            <td align="center">Intel GPU</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++</td>
        </tr>
        <tr>
            <td align="center">NVIDIA GPU</td>
            <td align="center">NVIDIA cuBLAS</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">LLVM*, hipSYCL</td>
        </tr>
        <tr>
            <td align="center">x86 CPU</td>
            <td align="center">NETLIB LAPACK</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*, hipSYCL</td>
        </tr>
	    <tr >
            <td align="center">AMD GPU</td>
            <td align="center">AMD rocBLAS </td>
            <td align="center">Dynamic, Static</td>
            <td align="center">hipSYCL</td>
        </tr>
        <tr>
            <td rowspan=3 align="center">LAPACK</td>
            <td align="center">x86 CPU</td>
            <td rowspan=2 align="center">Intel(R) oneAPI Math Kernel Library</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*</td>
        </tr>
        <tr>
            <td align="center">Intel GPU</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++</td>
        </tr>
        <tr>
            <td align="center">NVIDIA GPU</td>
            <td align="center">NVIDIA cuSOLVER</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">LLVM*</td>
        </tr>
        <tr>
            <td rowspan=3 align="center">RNG</td>
            <td align="center">x86 CPU</td>
            <td rowspan=2 align="center">Intel(R) oneAPI Math Kernel Library</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*</td>
        </tr>
        <tr>
            <td align="center">Intel GPU</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++</td>
        </tr>
        <tr>
            <td align="center">NVIDIA GPU</td>
            <td align="center">NVIDIA cuRAND</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">LLVM*</td>
        </tr>
    </tbody>
</table>

#### Windows*

<table>
    <thead>
        <tr align="center" >
            <th>Domain</th>
            <th>Backend</th>
            <th>Library</th>
            <th>Supported Link Type</th>
            <th>Supported Compiler</th>	
        </tr>
    </thead>
    <tbody>
        <tr>
            <td rowspan=3 align="center">BLAS</td>
            <td align="center">x86 CPU</td>
            <td rowspan=2 align="center">Intel(R) oneAPI Math Kernel Library</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*</td>
        </tr>
        <tr>
            <td align="center">Intel GPU</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++</td>
        </tr>
        <tr>
            <td align="center">x86 CPU</td>
            <td align="center">NETLIB LAPACK</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*</td>
        </tr>
        <tr>
            <td rowspan=2 align="center">LAPACK</td>
            <td align="center">x86 CPU</td>
            <td rowspan=2 align="center">Intel(R) oneAPI Math Kernel Library</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*</td>
        </tr>
        <tr>
            <td align="center">Intel GPU</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++</td>
        </tr>
        <tr>
            <td align="center">RNG</td>
            <td align="center">x86 CPU</td>
            <td align="center">Intel(R) oneAPI Math Kernel Library</td>
            <td align="center">Dynamic, Static</td>
            <td align="center">DPC++, LLVM*</td>
        </tr>
    </tbody>
</table>

\* LLVM - [Intel project for LLVM* technology](https://github.com/intel/llvm) with support for [NVIDIA CUDA](https://intel.github.io/llvm-docs/GetStartedGuide.html#build-dpc-toolchain-with-support-for-nvidia-cuda)
```
![Capture](https://user-images.githubusercontent.com/28574953/153960827-681b4f0d-5b87-48db-ba45-148a4ccc9194.JPG)

Reviewer 2: @sbalint98 I missed that we had these tables. I think these tables cover the three new tables we just added. If you agree, can you please remove those? Very sorry for the inconvenience. 

Also, for this table, can we remove the hipSYCL from BLAS/NVIDIA GPU/NVIDIA cuBLAS row? (not tested)

Reviewer 3: @mmeterel Can you elaborate on the rationale behind not just giving the hipSYCL cuBLAS backend a quick test run? It's not clear to me why we wouldn't advertise something that we know works and that we support. hipSYCL is often misunderstood as having a particular focus on AMD hardware, when it is in fact far more general and supports other hardware like NVIDIA GPUs just as well. We would like to prevent this impression.

Is this purely related to the organization of PRs and retaining a rocBLAS focus in this PR?

Reviewer 2: @illuhad There are mainly two reasons why I would prefer a two step approach for adding cuBLAS support for hipSYCL.

1. The first one is the organization of PRs as you mentioned. As the PR title says, this is about adding rocBLAS backend. 
2. Second, as you have seen in our offline discussions, correct testing of rocBLAS + hipSYCL locally took some time. For cuBLAS + hipSYCL, we would like to first validate it locally - which might take a while and currently don't have BW. We also need to discuss internally how/if we are going to support it in CI and make a plan for it. 

Otherwise, we are very excited to add hipSYCL support into our project and looking forward to extend its capabilities. Is there another PR for Intel GPUs coming soon as well? :)

Reviewer 3: @mmeterel Thank you for the explanation. I believe your existing hipSYCL installation might already be able to target CUDA. You'd only have to point cmake to CUDA when building hipSYCL if CUDA is installed in a non-standard location :-)

I think for Intel GPUs we will need some more time, as the hipSYCL level zero backend does not yet support backend interoperability to the same extent as the CUDA or HIP backends. So, oneMKL+hipSYCL+Intel GPUs is a bit further in the future, but it would be great to have support for this at some point :)

@sbalint98 Can you move the addition of hipSYCL cuBLAS to the documentation to a separate PR?

Author: @mkrainiuk @mmeterel Thank you for the changes, I applied them, let me know if there is still anything else to change. 

@illuhad I'll create a PR as soon as this is in ;)

Reviewer: Thanks @sbalint98! I think we no longer need https://github.com/sbalint98/oneMKL/tree/ustream-add-rocm#supported-configurations-with-hipsycl, could you please remove it?

Reviewer: And I think we discussed enabling hipSYCL in the future PR after full testing <html>
<body>
<!--StartFragment-->

BLAS | NVIDIA cuBLAS | Dynamic, Static | LLVM*<s>, hipSYCL</s>
-- | -- | -- | --

<!--EndFragment-->
</body>
</html>

@mmeterel please correct me if I'm wrong.

Reviewer 2: > We have not tested cuBLAS backend with hipSYCL locally yet. Could you save it for a future PR?

@sbalint98 I still see hipSYCL for cuBLAS row (unless I am looking at an old page). Could you please double check?

Author: my bad, sorry, now removed

- Accelerators
    - Intel(R) Processor Graphics GEN9
    - NVIDIA(R) TITAN RTX(TM) (Linux* only. cuRAND backend tested also with Quadro and A100 GPUs. Not tested with other NVIDIA GPU families and products.)

>     - ROCm enabled AMD GPUs see [here](https://github.com/RadeonOpenCompute/ROCm#hardware-and-software-support) tested on AMD Vega 20 (gfx906)

Reviewer 2: AMD(R) and TM? Could you please check how we should refer to company and product names?

- Accelerators
    - Intel(R) Processor Graphics GEN9
    - NVIDIA(R) TITAN RTX(TM) (Linux* only. cuRAND backend tested also with Quadro and A100 GPUs. Not tested with other NVIDIA GPU families and products.)

>     - ROCm enabled AMD GPUs see [here](https://github.com/RadeonOpenCompute/ROCm#hardware-and-software-support)

Reviewer 2: Could you please list the AMD GPU models tested?

    </tbody>
</table>

> #### For compiling with hipSYCL:

Reviewer: Why do we need hipSYCL note here? The requirement for hipSYCL with the link to it are listed below. As an option you can move it to "Selection of Compilers"

Author: That makes sense. Moving it over.

Author: Moved to "Selection of Compilers"

Reviewer: Thanks!

> see [here](https://github.com/illuhad/hipSYCL/blob/develop/doc/installing.md#software-dependencies)

Reviewer 2: ```suggestion
hipSYCL compiler
```

    </tbody>
</table>

#### For compiling with hipSYCL:

> General, and installation of hipSYCL and the required dependencies for the selected backends. see [here](https://github.com/illuhad/hipSYCL/blob/develop/doc/installing.md#software-dependencies)

Reviewer 2: I would recommend listing all software requirements in a table form here. This will help having a self contained repo here.

[Intel project for LLVM* technology binary for x86 CPU](https://github.com/intel/llvm/releases) | Daily builds (experimental) tested with [20200331](https://github.com/intel/llvm/releases/download/20200331/dpcpp-compiler.tar.gz) | No | *N/A* | *Installed by user* | [Apache License v2](https://github.com/intel/llvm/blob/sycl/sycl/LICENSE.TXT)
[Intel project for LLVM* technology source for NVIDIA GPU](https://github.com/intel/llvm/releases) | Daily source releases: tested with [20200421](https://github.com/intel/llvm/tree/20200421) | No | *N/A* | *Installed by user* | [Apache License v2](https://github.com/intel/llvm/blob/sycl/sycl/LICENSE.TXT)
[Intel(R) oneAPI Math Kernel Library](https://software.intel.com/en-us/oneapi/onemkl) | latest | Yes | apt | /opt/intel/inteloneapi/mkl | [Intel Simplified Software License](https://software.intel.com/en-us/license/intel-simplified-software-license)
[NVIDIA CUDA SDK](https://developer.nvidia.com/cublas) | 10.2 | No | *N/A* | *Installed by user* |[End User License Agreement](https://docs.nvidia.com/cuda/eula/index.html)

> [rocBLAS](https://rocmdocs.amd.com/en/latest/ROCm_Tools/rocblas.html) | 4.5 | No | *N/A* | *Installed by user* |[AMD License](https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/LICENSE.md)

Reviewer 2: ```suggestion
[AMD rocBLAS](https://rocmdocs.amd.com/en/latest/ROCm_Tools/rocblas.html) | 4.5 | No | *N/A* | *Installed by user* |[AMD License](https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/develop/LICENSE.md)
```

Reviewer 2: Is rocBLAS sufficient?

*Note: Conan package manager automates the process of getting required packages, so that you do not have to go to different web location and follow different instructions to install them.*

---
## Build Setup with HipSYCL

> 1. Make sure that the dependencies of hipSYCL are fulfilled. 

Reviewer 2: Please provide a link to dependencies.

*Note: Conan package manager automates the process of getting required packages, so that you do not have to go to different web location and follow different instructions to install them.*

---
## Build Setup with HipSYCL

1. Install hipSYCL with the prefered backends enabled see the [hipSYCL documentation](https://github.com/illuhad/hipSYCL/blob/develop/doc/installing.md) for instructions

> 2. Install an appropriate version of rocBLAS see instructions [here](https://rocblas.readthedocs.io/en/master/install.html)

Reviewer 2: Please list all the requirements needed. AMD HIP, AMD ROCm, AMD LLVM,  AMD rocBLAS.... Also, since we already mentioned the version number above, no need for "appropriate"

Reviewer 2: discussed offline.

-DENABLE_CURAND_BACKEND=True   \
```

### Building with hipSYCL
HipSYCL currently supports the BLAS domain with the following backends:
* NETLIB
* MKLCPU
* CUBLAS
* ROCBLAS

> From which the ROCBLAS backend is only available when compiling with hipSYCL

Reviewer 2: Can we add a section under "Build Options" for hipSYCL - and add these information there?

> #### Building oneMKL to target AMD GPU with hipSYCL

Reviewer 2: ```suggestion
#### Building for ROCm (with hipSYCL)
```

Reviewer 2: Let's follow the existing header styles

With the AMD rocBLAS backend:

- On Linux*

```bash
# Inside <path to onemkl>
mkdir build && cd build
cmake .. -DENABLE_CUBLAS_BACKEND=False                     \

>          -DENABLE_MKLCPU_BACKEND=False/True                \   

Reviewer 2: ```suggestion
         -DENABLE_MKLCPU_BACKEND=False/True                \   # hipSYCL supports MKLCPU backend     
```

>          -DENABLE_NETLIB_BACKEND=False/True                \   # hipSYCL supports both NETLIB and MKLCPU backends

Reviewer 2: ```suggestion
         -DENABLE_NETLIB_BACKEND=False/True                \   # hipSYCL supports NETLIB backend
```

         -DENABLE_MKLGPU_BACKEND=False                     \   # disable Intel MKL GPU backend
         -DENABLE_ROCBLAS_BACKEND=True                     \
         -DTARGET_DOMAINS=blas                             \   # hipSYCL only supports the BLAS domain
         -DHIPSYCL_TARGETS=omp\;hip:gfx906                 \   # Specify the targetted device architectures 
         -DONEMKL_SYCL_IMPLEMENTATION=hipSYCL              \   # Use the hipSYCL cmake integration
         [-DREF_BLAS_ROOT=<reference_blas_install_prefix>] \   # required only for testing
cmake --build .
ctest
cmake --install . --prefix <path_to_install_dir>
```

**AMD GPU device architectures**  

The device architecture can be retrieved via the `rocminfo` tool. The architecture will be displayed in the `Name:` row.

A few often used architectures are listed below:
| architecture | AMD GPU name |
| ----         | ----         |

> | gfx906       | Radeon Instinct MI50/60 <br> Radeon (Pro) VII |

Reviewer 2: Please make sure the product names are listed appropriately as recommended by companies.

Author: Used the official AMD websites of each product for reference

| gfx908       | Radeon Instinct MI 100 |
| gfx900       | Radeon Instinct MI 25 <br> Radeon RX Vega 64/56 |

> #### Building oneMKL to target CUDA GPU with hipSYCL  

Reviewer 2: We have not tested this functionality yet (and haven't discussed the coverage). Can we submit this as a separate PR?

> OneMKL compiled with hipSYCL can also target CUDA GPUs. With the following modifications to the above configuration settings, the cuBLAS domain can be enabled:

Reviewer 2: ```suggestion
oneMKL compiled with hipSYCL can also target Nvidia GPUs. With the following modifications to the above configuration settings, cuBLAS domain can be enabled:
```

```blas
-DENABLE_ROCBLAS_BACKEND=False
-DENABLE_CUBLAS_BACKEND=True

> -DHIPSYCL_TARGETS=omp\;cuda:sm_61  # sm_61 for targetting gtx 1080 TI GPPU

Reviewer 2: ```suggestion
-DHIPSYCL_TARGETS=omp\;cuda:sm_61  # sm_61 for targeting gtx 1080 TI GPU
```

Conan Option | CMake Option | Supported Values | Default Value 
 :---------- | :----------- | :--------------- | :---          
build_shared_libs        | BUILD_SHARED_LIBS        | True, False         | True      
enable_mklcpu_backend    | ENABLE_MKLCPU_BACKEND    | True, False         | True      
enable_mklgpu_backend    | ENABLE_MKLGPU_BACKEND    | True, False         | True      
*Not Supported*          | ENABLE_CUBLAS_BACKEND    | True, False         | False     
*Not Supported*          | ENABLE_CUSOLVER_BACKEND  | True, False         | False     
*Not Supported*          | ENABLE_CURAND_BACKEND    | True, False         | False     
*Not Supported*          | ENABLE_NETLIB_BACKEND    | True, False         | False     
*Not Supported*          | ENABLE_ROCBLAS_BACKEND   | True, False         | False     
enable_mklcpu_thread_tbb | ENABLE_MKLCPU_THREAD_TBB | True, False         | True      
build_functional_tests   | BUILD_FUNCTIONAL_TESTS   | True, False         | True      
build_doc                | BUILD_DOC                | True, False         | False     
target_domains (list)    | TARGET_DOMAINS (list)    | blas, lapack, rng   | All domains 

> 

Reviewer 2: Please add a simple sentence that summarizes the below tables.

**Backend Support of Compilers in the BLAS Domain**

> Compiler | mklcpu | netlib | mklgpu | cublas | rocblas

Reviewer 2: for order: mklcpu mklgpu netlib cublas rocblas (match other domains)

-----    | ----   | ------ | ----   | ---    | ---
dpc++    | Yes    | Yes    | Yes    | No     | No
LLVM     | Yes    | Yes    | Yes    | Yes    | No
hipSYCL  | Yes    | Yes    | No     | No     | Yes

> **Backend Support of Compilers in the lapack Domain**

Reviewer 2: ```suggestion
**Backend Support of Compilers in the LAPACK Domain**
```

Compiler | mklcpu | mklgpu | cusolver 
-----    | ----   | ------ | ---      
dpc++    | Yes    | Yes    | No       
LLVM     | Yes    | Yes    | Yes      
hipSYCL  | No     | No     | No       

> **Backend Support of Compilers in the rng Domain**

Reviewer 2: ```suggestion
**Backend Support of Compilers in the RNG Domain**
```

Conan Option | CMake Option | Supported Values | Default Value 
 :---------- | :----------- | :--------------- | :---          
build_shared_libs        | BUILD_SHARED_LIBS        | True, False         | True      
enable_mklcpu_backend    | ENABLE_MKLCPU_BACKEND    | True, False         | True      
enable_mklgpu_backend    | ENABLE_MKLGPU_BACKEND    | True, False         | True      
*Not Supported*          | ENABLE_CUBLAS_BACKEND    | True, False         | False     
*Not Supported*          | ENABLE_CUSOLVER_BACKEND  | True, False         | False     
*Not Supported*          | ENABLE_CURAND_BACKEND    | True, False         | False     
*Not Supported*          | ENABLE_NETLIB_BACKEND    | True, False         | False     
*Not Supported*          | ENABLE_ROCBLAS_BACKEND   | True, False         | False     
enable_mklcpu_thread_tbb | ENABLE_MKLCPU_THREAD_TBB | True, False         | True      
build_functional_tests   | BUILD_FUNCTIONAL_TESTS   | True, False         | True      
build_doc                | BUILD_DOC                | True, False         | False     
target_domains (list)    | TARGET_DOMAINS (list)    | blas, lapack, rng   | All domains 

> The following tables summarize the supported backends for each domain and compiler combination:

Reviewer 2: ```suggestion
The following tables summarize the supported compilers for each domain and backend combination:
```

**Backend Support of Compilers in the BLAS Domain**

Compiler | mklcpu | mklgpu | netlib | cublas | rocblas
-----    | ----   | ------ | ----   | ---    | ---
dpc++    | Yes    | Yes    | Yes    | No     | No
LLVM     | Yes    | Yes    | Yes    | Yes    | No
hipSYCL  | Yes    | No     | Yes    | No     | Yes

**Backend Support of Compilers in the LAPACK Domain**

Compiler | mklcpu | mklgpu | cusolver 
-----    | ----   | ------ | ---      
dpc++    | Yes    | Yes    | No       
LLVM     | Yes    | Yes    | Yes      
hipSYCL  | No     | No     | No       

**Backend Support of Compilers in the RNG Domain**

Compiler | mklcpu | mklgpu | curand 
-----    | ----   | ------ | ---    
dpc++    | Yes    | Yes    | No     
LLVM     | Yes    | Yes    | Yes    
hipSYCL  | No     | No     | No     

*Note: `build_functional_tests` and related CMake option affects all domains at a global scope.*

> *Note: When building with hipSYCL, `-DHIPSYCL_TARGETS` additionally needs to be provided according to the targeted hardware. For the options, see the tables in the hipSYCL specific sections*

Reviewer: Normalize the font

```suggestion
*Note: When building with hipSYCL, `-DHIPSYCL_TARGETS` additionally needs to be provided according to the targeted hardware. For the options, see the tables in the hipSYCL specific sections*

```

-DENABLE_CURAND_BACKEND=True   \
```

### Building for rocBLAS with hipSYCL

> with the rocBLAS backend:

Reviewer 2: Specify Linux only

> with the rocBLAS backend:

Reviewer 2: ```suggestion
With the rocBLAS backend:
```

```bash
# Inside <path to onemkl>
mkdir build && cd build
cmake .. -DENABLE_CUBLAS_BACKEND=False                     \

>          -DENABLE_MKLCPU_BACKEND=False                     \   # disable Intel MKL CPU backend

Reviewer 2: Do we need this? hipSYCL is supporting Intel MKL CPU backend, right?
Similarly, how about ENABLE_NETLIB_BACKEND? It is supported by hipSYCL, right?

         -DENABLE_MKLGPU_BACKEND=False                     \   # disable Intel MKL GPU backend
         -DENABLE_ROCBLAS_BACKEND=True                     \
         -DTARGET_DOMAINS=blas                             \   # hipSYCL only supports the BLAS domain

>          -DHIPSYCL_TARGETS=omp\;hip:gfx906                 \   # Specify the targetted device architectures

Reviewer 2: Please specify options for different architectures (Vega, MI)

*Not Supported*          | ENABLE_CUBLAS_BACKEND    | True, False         | False
*Not Supported*          | ENABLE_CURAND_BACKEND    | True, False         | False
*Not Supported*          | ENABLE_NETLIB_BACKEND    | True, False         | False
*Not Supported*          | ENABLE_ROCBLAS_BACKEND   | True, False         | False

> *Not Supported*          | ONEMKL_SYCL_IMPLEMENTATION| dpc++, hipsycl     | dpc++

Reviewer: `ONEMKL_SYCL_IMPLEMENTATION| dpc++, hipsycl` is very confusing since `hipsycl` has its own build options below. Is it possible to name these two tables as `ONEMKL_SYCL_IMPLEMENTATION=dpc++(default)` and `ONEMKL_SYCL_IMPLEMENTATION=hipsycl` and remove this line?

Author: Done, also added a bit of explanation.

In the following table, the supported options for each type of SYCL implementation are listed.

All options specified in the Conan section are available to CMake. You can specify these options using `-D<cmake_option>=<value>`.

The following table provides a detailed mapping of options between Conan and CMake.

Conan Option | CMake Option | Supported Values | Default Value | Supported Compilers
 :---------- | :----------- | :--------------- | :---          | :---
build_shared_libs        | BUILD_SHARED_LIBS        | True, False         | True      | All
enable_mklcpu_backend    | ENABLE_MKLCPU_BACKEND    | True, False         | True      | All
enable_mklgpu_backend    | ENABLE_MKLGPU_BACKEND    | True, False         | True      | All
*Not Supported*          | ENABLE_CUBLAS_BACKEND    | True, False         | False     | LLVM
*Not Supported*          | ENABLE_CUSOLVER_BACKEND  | True, False         | False     | LLVM
*Not Supported*          | ENABLE_CURAND_BACKEND    | True, False         | False     | LLVM
*Not Supported*          | ENABLE_NETLIB_BACKEND    | True, False         | False     | All

> *Not Supported*          | ENABLE_ROCBLAS_BACKEND   | False               | False     | hipSYCL

Reviewer 2: Supported Values should be True, False, right?

Author: yes...

enable_mklcpu_thread_tbb | ENABLE_MKLCPU_THREAD_TBB | True, False         | True      | All
build_functional_tests   | BUILD_FUNCTIONAL_TESTS   | True, False         | True      | All
build_doc                | BUILD_DOC                | True, False         | False     | All
target_domains (list)    | TARGET_DOMAINS (list)    | blas, lapack, rng   | All domains | see table below

> **Domain Support of Compilers**

Reviewer 2: This table is little misleading. Not all backends are supported by all compilers. 
I would suggest splitting this table into three, one for each domain (blas/lapack/rng). 
And then for each table, show which backend is supported by which compiler. 
We can deduce this information from the above table, but putting it by domain/backend will make it very clear imho.

Reviewer 2: If we add these tables for each domain, we don't need to add Supported Compilers column to "Conan vs Cmake" table. The purpose of that table is to differentiate Conan vs. Cmake options.

Author: I agree that makes sense... I have created the domain-specific tables. Please check, that I got everything right

*Not Supported*          | ENABLE_CUBLAS_BACKEND    | True, False         | False
*Not Supported*          | ENABLE_CURAND_BACKEND    | True, False         | False
*Not Supported*          | ENABLE_NETLIB_BACKEND    | True, False         | False
*Not Supported*          | ENABLE_ROCBLAS_BACKEND   | True, False         | False
*Not Supported*          | ONEMKL_SYCL_IMPLEMENTATION| dpc++, hipsycl     | dpc++
enable_mklcpu_thread_tbb | ENABLE_MKLCPU_THREAD_TBB | True, False         | True
build_functional_tests   | BUILD_FUNCTIONAL_TESTS   | True, False         | True
build_doc                | BUILD_DOC                | True, False         | False
target_domains (list)    | TARGET_DOMAINS (list)    | blas, lapack, rng   | All domains

*Note: `build_functional_tests` and related CMake option affects all domains at a global scope.*

> *Note: When building with hipSYCL, the `-DHIPSYCL_TARGETS` additionally needs to be provided according to the targetted hardware. for the options, see the tables in the hipSYCL specific sections*

Reviewer 2: ```suggestion
*Note: When building with hipSYCL, `-DHIPSYCL_TARGETS` additionally needs to be provided according to the targeted hardware. For the options, see the tables in the hipSYCL specific sections*
```
